In [16]:
def sqsum(k,m):
    total = 0
    for i in range(k-m,k+1):
        total += i**2
    return total

def right(total,m):
    n = (total/m)**0.5 + m/2
    n = int(n)
    t = sqsum(n,m-1)
    if t == total:
        return n-m+1
    if t > total:
        while t > total:
            n -= 1
            t = sqsum(n,m-1)
            if t == total:
                return n-m+1
        return -1
    while t < total:
        n += 1
        t = sqsum(n,m-1)
        if t == total:
            return n-m+1
    return -1

    

sqsum(24,4)

2430

In [17]:
right(2430,4)

-1

上は単純に機械的に計算した。右辺と左辺をキャンセルすることで、もっと良い計算ができると思う。
$$(k-m)^2+\cdots+k^2=(n+1)^2+\cdots+(n+m)^2$$
$$(n+m)^2-(k-m)^2+\cdots+(n+1)^2-(k-1)^2=k^2$$
$$(n+k)(n-k+2m)+\cdots+(n+k)(n-k+2)=k^2$$
つまり、$k^2$は$n+k$の倍数である。
$$2+4+\cdots+2m=m(m+1)$$
$$(n+k)m(n-k+m+1)=k^2$$
$m$の倍数でもある。

$k$と$m$を決めると、$n$は一意に決まる。

1. $k^2$は$m$で割り切れなければいけない。
2. $D=4k^2 - 4km - 4k + 4k^2/m + m^2 + 2m + 1$が平方数でなければいけない。
3. $n=(-m+\sqrt{D}-1)/2$である。

In [23]:
24**2/3/48

4.0

In [22]:
(24-24+4)

4

In [30]:
import sympy
n,k,m,km = sympy.symbols("n k m km")
sympy.solve((n+k)*(n-k+m+1)-km,n)[1].simplify()*2

-m + sqrt(4*k**2 - 4*k*m - 4*k + 4*km + m**2 + 2*m + 1) - 1

In [31]:
(4*k**2 - 4*k*m - 4*k + 4*km + m**2 + 2*m + 1).simplify()

4*k**2 - 4*k*m - 4*k + 4*km + m**2 + 2*m + 1

In [14]:
def issq(n):
    m = n**0.5
    return m == int(m)

def pe261(k):
    for m in range(1,k):
        if k**2 % m != 0:
            continue
        km=k**2/m
        D = 4*k**2 - 4*k*m - 4*k + 4*km + m**2 + 2*m + 1
        n = (-m+D**0.5-1)/2
        if n < k:
            break
        if not issq(D):
            continue
        yield m,n

for k in range(1,10000):
    for m,n in pe261(k):
        print(k,m,n)


break
break
4 1 4.0
break
break
break
break
break
12 2 12.0
break
break
break
break
break
break
21 1 28.0
break
break
24 3 24.0
break
break
break
break
break
break
break
break
break
break
break
break
break
40 4 40.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
60 5 60.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
84 6 84.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
110 2 132.0
break
break
112 7 112.0
break
break
break
break
break
break
break
120 1 168.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
144 8 144.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
180 9 180.0
break
break
break
break
break
break
br

break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
1860 30 1860.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
1984 31 1984.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
brea

break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
3612 42 3612.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
b

break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
5304 51 5304.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
b

break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
7080 59 7080.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
b

break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
8844 66 8844.0
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
break
b

分枝を見付ける問題のようだ。

In [67]:
18240-17860

380

In [68]:
17860-17484

376

上の数列は、いくつかの法則がある。だから、法則を知れば、全部を列挙する必要はなくなる。

最初の系列は、

    4 1 4
    12 2 12
    24 3 24
    40 4 40

差が8、12、16と増えていく。
次の系列は

    21 1 28     7x3  7x4     8
    110 2 132  22x5 22x6    24
    315 3 360  45x7 45x8    48
    684 4 760  76x9 76x10   80
   
差は89、205、369。一見するとどういう規則かわからない。

    120 1 168      24x5 24x7      1
    1080 2 1320   120x9 120x11    5
    4368 3 5040   336x13 336x15  14
    
なんとか見えてきた。
m=1の解を見付けられれば、それから派生してm=2やm=3の解が決まるはず。    

系列の初項を見付ける方法と、漸化式が必要。


### 初項

初項は$m=1$の場合、つまり左辺に2乗項が2つ、右辺に1つの、ピタゴラス型の式で、しかも左辺の2項が1違いの場合。
$$(n+k)(n-k+2)=k^2$$
となるようなケースである。これは解ける。

$$n^2+2n+2k(1-k)=0$$

$k$に対し、$n=\sqrt{2k^2-2k+1}-1$である。


In [8]:
from sympy import *
init_printing(use_latex='mathjax')
n,k = symbols("n k")
solve((n+k)*(n-k+2)-k**2, n)
#((n+k)*(n-k+2)-k**2).expand()

⎡     ________________         ________________    ⎤
⎢    ╱    2                   ╱    2               ⎥
⎣- ╲╱  2⋅k  - 2⋅k + 1  - 1, ╲╱  2⋅k  - 2⋅k + 1  - 1⎦

In [13]:
for k in range(1,10000):
    D = 2*k**2-2*k+1
    if issq(D):
        n = D**0.5-1
        print(k,n)

1 0.0
4 4.0
21 28.0
120 168.0
697 984.0
4060 5740.0


上のようなループで処理をしていては時間がかかりすぎるので、`if`をみたす$k$だけを直接算出する方法はないか。

    1x4   1x4
    3x7   4x7
    5x24  7x24
    17x41 24x41
    29x140, 41x140

$n$のほうが規則性がわかりやすい。ただし、$1,4,7,24,41,140$の規則性がわからない。

In [17]:
gcd(4060,5740),4060/140, 5740/140

(140, 29.0, 41.0)

### 漸化式
漸化式のほうはどうなるんだろう。

一見しただけでは、$m=1$のケースから$m=2$の場合の$n$や$k$を推定できない気はする。

漸化式ではなく、